stdlib/immutablearray.gr

```diff
@@ -0,0 +1,921 @@
+  if (numNotAppended > 0) {
+    let appendLen = Array.length(toAppend)
+    let newTail = Array.slice(appendLen - numNotAppended, appendLen, toAppend)
+    replaceTail(newTail, newArray)
+  } else {
+    newArray
+  }
+}
+
+// a helper data structure used for building an array piece by piece
+record Builder<a> {
+  tail: Array<a>,
+  nodes: List<Node<a>>,
+  numNodes: Number,
+}
+
+// Flatten an array into a builder
+let arrayToBuilder = array => {
+  let rec reduceFn = (acc, node) => {
+    match (node) {
+      Tree(subTree) => Array.reduce(reduceFn, acc, subTree),
+      Leaf(_) => [node, ...acc],
+    }
+  }
+  let { tail, root, length, _ } = array
+  {
+    tail,
+    nodes: Array.reduce(reduceFn, [], root),
+    numNodes: length >> branchingBits,
+  }
+}
+
+// For use to compress a large (> 32) list of nodes into subtrees
+let rec compressNodes = (nodes, acc) => {
+  let node = Array.fromList(List.take(branchingFactor, nodes))
+  let remaining = List.drop(branchingFactor, nodes)
+  let newAcc = [Tree(node), ...acc]
+  match (remaining) {
+    [] => List.reverse(newAcc),
+    _ => compressNodes(remaining, newAcc),
+  }
+}
+
+let builderToArray = builder => {
+  // Builds the non-tail portion of an array
+  let rec buildTree = (nodes, numNodes) => {
+    let newNodeSize = Number.ceil(numNodes / branchingFactor)
+    match (newNodeSize) {
+      1 => Array.fromList(nodes),
+      _ => buildTree(compressNodes(nodes, []), newNodeSize),
+    }
+  }
+
+  let { tail, nodes, numNodes } = builder
+  match (numNodes) {
+    0 => { length: Array.length(tail), shift: branchingBits, root: [>], tail },
+    _ => {
+      let treeSize = numNodes * branchingFactor
+      let depth = log32floor(treeSize - 1)
+      {
+        length: treeSize + Array.length(tail),
+        shift: Number.max(1, depth) * branchingBits,
+        root: buildTree(nodes, numNodes),
+        tail,
+      }
+    },
+  }
+}
+
+// Append a chunk of <= 32 values to a builder
+let appendBuilder = (toAppend, builder) => {
+  let { tail, nodes, numNodes } = builder
+  let (appended, numNotAppended) = arrayAppendMax32(tail, toAppend)
+
+  if (numNotAppended >= 0) {
+    let appendLen = Array.length(toAppend)
+    {
+      tail: Array.slice(appendLen - numNotAppended, appendLen, toAppend),
+      nodes: [Leaf(appended), ...nodes],
+      numNodes: numNodes + 1,
+    }
+  } else {
+    { tail: appended, nodes, numNodes }
+  }
+}
+
+/**
+ * Creates a new array with the elements of the first array followed by
+ * the elements of the second array.
+ *
+ * @param array1: The array containing elements to appear first
+ * @param array2: The array containing elements to appear second
+ * @returns The new array containing elements from `array1` followed by elements from `array2`
+ *
+ * @example append(fromList([1, 2]), fromList([3, 4, 5])) == fromList([1, 2, 3, 4, 5])
+ *
+ * @since v0.6.0
+ */
+export let append = (array1, array2) => {
+  // Magic number of 4 determined best from benchmarks
```

Comment:
Benchmarks in grain or is this like a comment from elm?

Reply:
It's a comment from elm; I can make this clear in the comment